remote_mapper.go

```diff
@@ -0,0 +1,196 @@
+package influxdb
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/influxdb/influxdb/influxql"
+)
+
+const (
+	MAX_MAP_RESPONSE_SIZE = 1024 * 1024
+)
+
+// RemoteMapper implements the influxql.Mapper interface. The engine uses the remote mapper
+// to pull map results from shards that only exist on other servers in the cluster.
+type RemoteMapper struct {
+	dataNodes []*DataNode
+	resp      *http.Response
+	results   chan interface{}
+	unmarshal influxql.UnmarshalFunc
+	complete  bool
+
+	Call            string   `json:",omitempty"`
+	Database        string   `json:",omitempty"`
+	MeasurementName string   `json:",omitempty"`
+	TMin            int64    `json:",omitempty"`
+	TMax            int64    `json:",omitempty"`
+	SeriesIDs       []uint64 `json:",omitempty"`
+	ShardID         uint64   `json:",omitempty"`
+	Filters         []string `json:",omitempty"`
+	WhereFields     []*Field `json:",omitempty"`
+	SelectFields    []*Field `json:",omitempty"`
+	SelectTags      []string `json:",omitempty"`
+	Limit           int      `json:",omitempty"`
+	Offset          int      `json:",omitempty"`
+	Interval        int64    `json:",omitempty"`
+	ChunkSize       int      `json:",omitempty"`
+}
+
+// Responses get streamed back to the remote mapper from the remote machine that runs a local mapper
+type MapResponse struct {
+	Err       string `json:",omitempty"`
+	Data      []byte
+	Completed bool `json:",omitempty"`
+}
+
+// Open is a no op, real work is done starting with Being
+func (m *RemoteMapper) Open() error { return nil }
+
+// Close the response body
+func (m *RemoteMapper) Close() {
+	if m.resp != nil && m.resp.Body != nil {
+		m.resp.Body.Close()
+	}
+}
+
+// Begin sends a request to the remote server to start streaming map results
+func (m *RemoteMapper) Begin(c *influxql.Call, startingTime int64, chunkSize int) error {
+	// get the function for unmarshaling results
+	f, err := influxql.InitializeUnmarshaller(c)
+	if err != nil {
+		return err
+	}
+	m.unmarshal = f
+
+	if c != nil {
+		m.Call = c.String()
+	}
+	m.ChunkSize = chunkSize
+	m.TMin = startingTime
+
+	// send the request to map to the remote server
+	b, err := json.Marshal(m)
+	if err != nil {
+		return err
+	}
+
+	// request to start streaming results
+	resp, err := http.Post(m.dataNodes[0].URL.String()+"/run_mapper", "application/json", bytes.NewReader(b))
```

Comment:
Should this be `m.dataNode[rand.Intn(len(m.dataNodes))]`to avoid a single node handling all the traffic?  Or is `m.dataNodes` already randomized beforehand?